### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by symroe**
Couple of things in this PR:

## Party Choices

The old party choices method was _really_ complex, both in the code and in the query. It was basically trying to be smart about presenting a useful list of parties to the user in various places. It did this by:

* Checking party registration status
* Ordering by current memberships
* Optionally ordering by total memberships
* Including parties with no memberships
* Optionally including descriptions

That's a lot for one method. It also looked at a threshold of candidates, and if the total number was under it, it would return a simple list of all parties.

That last point is important, because in the tests this method was used most of the time.

The new change:

* Adds fields to the `Party` model for "current" candidates and total candidates.
* Adds a management command for updating both of those figures
* Updates the party choices method to use them
* Adds a couple of helper methods
* Remove the number of candidates from the name. This is because the figures wont update in real time any more, and they might not be very helpful.

The "bulk add from SOPN" page used to load in 249.38 ms on my laptop. It's now down to 55.25 ms (and the memory footprint will be much smaller too).

## Lock Suggestions

This is a bit of a different change, and could be it's own PR, but I did it as part of looking at the performance of the bulk add process. 

In short:

* Clean up the suggested locks after locking
* Turn the context processor query in to a straight count of suggested locks
* Run a data migration to clean up all the locks for non-current elections
* Check for lock permissions before adding the numbers to the context

### CODE REVIEW

**Comment by symroe** [comment]
This pull request **introduces 2 alerts** when merging b3db5f6060ee994870d07d48515a9a7fc7898727 into 86893ae0e695fd3683e47d7ef97883a366954346 - [view on LGTM.com](https://lgtm.com/projects/g/DemocracyClub/yournextrepresentative/rev/pr-7f3834d898c073db28fba299acf1feafedb1c34c)

**new alerts:**

* 2 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by symroe** [comment]
This pull request **fixes 1 alert** when merging 80930288262788740c681556516767dc93fdf1de into 86893ae0e695fd3683e47d7ef97883a366954346 - [view on LGTM.com](https://lgtm.com/projects/g/DemocracyClub/yournextrepresentative/rev/pr-41cdcd8de3986fb46d14cd4fd49dda197c824027)

**fixed alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by chris48s** [comment]
This seems to work fine in the party drop-down on the person create and inline person add, but if I check this branch out, the drop-downs on the 2 bulk adding forms don't load any parties by default:

GIFrecord_2019-03-21_191007

GIFrecord_2019-03-21_191034


If I check out master, both of those 2 drop-downs do load parties by default.

File: ynr/apps/parties/tests/test_party_choices.py [hunk 1]
```diff
@@ -45,16 +45,14 @@ def test_enough_candidates_in_current_election(self):
...
             party_choices,
             [
                 (u"", u""),
-                (
-                    str(self.ld_party.ec_id),
-                    u"Liberal Democrats (30 candidates)",
-                ),
-                (str(self.green_party.ec_id), u"Green Party (15 candidates)"),
+                (str(self.ld_party.ec_id), u"Liberal Democrats"),
+                (str(self.green_party.ec_id), u"Green Party"),
```
Comment (chris48s) [inline ynr/apps/parties/tests/test_party_choices.py]: It looks like these tests are written using `str()`, `u""` for python 2/3 compatibility. While you're modifying these files, can we de-clutter some of this to just use strings so its a bit more readable?

File: ynr/apps/elections/views.py [hunk 2]
```diff
@@ -977,6 +977,9 @@ def form_valid(self, form):
             if lock:
                 suffix = "-lock"
                 pp = "Locked"
+                # If we're locking this, then the suggested posts
+                # can be deleted
+                pee.suggestedpostlock_set.all().delete()
```
Comment (chris48s) [inline ynr/apps/elections/views.py]: Can we wrap this function in a transaction so that if something throws an exception later in the function we roll this `delete()` back

File: ynr/context_processors.py [hunk 3]
```diff
@@ -57,12 +57,13 @@ def add_notification_data(request):
...
     result = {}
     if request.user.is_authenticated:
-        result["photos_for_review"] = QueuedImage.objects.filter(
-            decision="undecided"
-        ).count()
-        result["suggestions_to_lock"] = SuggestedPostLock.objects.filter(
-            postextraelection__candidates_locked=False
-        ).count()
+        groups = set(request.user.groups.values_list("name", flat=True))
+        if PHOTO_REVIEWERS_GROUP_NAME in groups:
```
Comment (chris48s) [inline ynr/context_processors.py]: This is all useful stuff being changed in this file, but the changes to this file aren't really related to the commit message. Just for the sake of future code archaeology, can we split this into 2 commits with descriptive messages.

**Comment by symroe** [comment]
[YNR performance optimisations ahead of SOPN parties](https://trello.com/c/ETRrd1gU/1700-ynr-performance-optimisations-ahead-of-sopn-parties)

**Comment by symroe** [comment]
> the 2 bulk adding forms don't load any parties by default

I think this is because you've not run the `parties_update_current_candidates` command?

As this would be the default for anyone new to the project, I've added a migration that adds the counts for an existing database.

I've also called the `parties_update_current_candidates` command at the end of the live site importer, so anyone populating the database for the first time will get useful values in the drop downs. 

As for why this happens on the bulk adding page and not the person add page, that's because the value of `include_non_current` is different. We only include parties with "current" candidates on the bulk add pages to save on page size – the bulk of candidates are one of ~10 parties, and we load that drop down 15 times per page. This doesn't matter so much on the single person edit page, so we load them all.

File: ynr/apps/elections/views.py (hunk 2 above)
Comment (symroe) [reply ynr/apps/elections/views.py]: It's already in one, https://github.com/DemocracyClub/yournextrepresentative/blob/80930288262788740c681556516767dc93fdf1de/ynr/apps/elections/views.py#L966-L968

**Comment by chris48s** [comment]
> I think this is because you've not run the `parties_update_current_candidates` command?

Nope. From last night's bash_history:

```
./manage.py migrate
./manage.py parties_update_current_candidates
./manage.py runserver
```

It is useful to populate in the migration/sync script though. :+1: 

> As for why this happens on the bulk adding page and not the person add page, that's because the value of include_non_current is different.

Right. Looking at my parties table, I've got values in the `total_candidates` column, but `current_candidates` is zero for every party. Not sure if that's because there's an issue with the `update_current_candidates` or just because my local DB is unrepresentative, but if the person create and inline person add forms look at  `total_candidates` and the bulk add ones look at `current_candidates` that would probably explain why it works in some cases but not in others. Is there any point in having some kind of default in this case, or is it sufficiently unlikely/esoteric that we don't care?

File: ynr/apps/sopn_parsing/management/commands/sopn_parsing_parse_tables.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
+import re
```
Comment (chris48s) [inline ynr/apps/sopn_parsing/management/commands/sopn_parsing_parse_tables.py]: I don't think this file belong in this commit/PR

**Comment by symroe** [comment]
Right, I see what you mean about your DB state. 

Of course I could add a check to the method to see if the queryset is empty, and fall back to showing all parties if it were, but that would add an additional query to the method to satisfy a case that is very unlikely to happen in real use.

Maybe this is something we can toggle with a setting, on by default, so local dev copies work well on a clean checkout?

Just to check, if you press "load more parties", you get the full list, right?

**Comment by symroe** [comment]
Another option – and a bit of a hack – is that if there are no candidates in current elections at all, I can set the current count to the total count, in the management command. This would never happen in real use, but would make dev copies work better.

**Comment by chris48s** [comment]
The main thing we're trying to do here is improve performance in production, so lets not compromise that. Realistically it is only really me and you running this in dev. As long as `update_current_candidates` will populate `current_candidates` with sensible numbers when we deploy it, I don't think there is a problem. We can punt "make parties drop-down work better on local dev installs" to an issue if there isn't an obvious fallback for this uncommon case.

**Comment by symroe** [comment]
This pull request **fixes 1 alert** when merging 69293cef012cd0a74f9c76ab0cc05f27fb9e77e3 into af5aeda6f20cd7e8ec6f26839ce3e9a2fe8c87e7 - [view on LGTM.com](https://lgtm.com/projects/g/DemocracyClub/yournextrepresentative/rev/pr-02dfe1c68fed6f9b8f66faf62f693faf47b56f55)

**fixed alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by symroe** [comment]
This pull request **fixes 1 alert** when merging 3f4d22c396bda2e60bdf8780219f3766921274ab into 4b74ed985352bc8c1ddc790e300ac9abd57e4608 - [view on LGTM.com](https://lgtm.com/projects/g/DemocracyClub/yournextrepresentative/rev/pr-73f3600c42ca4389dfa0c0f74d92dac86c4270d4)

**fixed alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*
